src/poetry/mixology/version_solver.py

```diff
@@ -39,36 +47,65 @@ class DependencyCache:
 
     def __init__(self, provider: Provider) -> None:
         self.provider = provider
-        self.cache: dict[
-            tuple[str, str | None, str | None, str | None, str | None],
-            list[DependencyPackage],
-        ] = {}
 
-        self.search_for = functools.lru_cache(maxsize=128)(self._search_for)
+        # self.cache maps a package name to a stack of cached package lists,
+        # ordered by the decision level which added them to the cache. This is
+        # done so that when backtracking we can maintain cache entries from
+        # previous decision levels, while clearing cache entries from only the
+        # rolled back levels.
+        #
+        # In order to maintain the integrity of the cache, `clear_level()`
+        # needs to be called in descending order as decision levels are
+        # backtracked so that the correct items can be popped from the stack.
+        self.cache: dict[DependencyCacheKey, list[list[DependencyPackage]]] = (
+            collections.defaultdict(list)
+        )
+        self.cached_dependencies_by_level: dict[int, list[DependencyCacheKey]] = (
```

Comment:
```suggestion
        self._cached_dependencies_by_level: dict[int, list[DependencyCacheKey]] = (
```

Maybe, we can add a leading `"_"` for internal attributes to clarify the interface of `DependencyCache`. (I'm aware that no attention was paid to this before.)